# AI-Powered Healthcare Diagnosis Assistant

## Data Preprocessing

### Project Objective

The objective of this notebook is to preprocess the validated healthcare dataset before model training. The preprocessing pipeline includes quality verification, rare class handling, feature preparation, label encoding, train-test splitting, and exporting cleaned datasets for machine learning models.

**Author:** Khushi Krishna

**Notebook:** 02_Data_Preprocessing.ipynb

In [1]:
# ============================================================
# AI-Powered Healthcare Diagnosis Assistant
# Notebook 02: Data Preprocessing
# Step 1: Import Required Libraries
# ============================================================

import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder

pd.set_option('display.max_columns', None)
pd.set_option('display.width', 1000)

print("Libraries imported successfully.")

Libraries imported successfully.


In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
dataset_path = "/content/drive/MyDrive/AI-Powered-Healthcare-Diagnosis-Assistant/Dataset/Final_Augmented_dataset_Diseases_and_Symptoms.csv"

df = pd.read_csv(dataset_path)

print("Dataset Loaded Successfully")
print(df.shape)

Dataset Loaded Successfully
(246945, 378)


In [4]:
# ============================================================
# Step 5: Create Original Dataset Backup
# ============================================================

original_df = df.copy()

print("Backup created successfully.")
print("Original Dataset Shape:", original_df.shape)

Backup created successfully.
Original Dataset Shape: (246945, 378)


In [5]:
# ============================================================
# Step 6: Compare Different Rare Disease Thresholds
# ============================================================

disease_counts = original_df["diseases"].value_counts()

thresholds = [5, 10, 20]

comparison = []

for threshold in thresholds:
    valid = disease_counts[disease_counts >= threshold].index
    filtered = original_df[original_df["diseases"].isin(valid)]

    comparison.append({
        "Threshold": threshold,
        "Remaining Diseases": filtered["diseases"].nunique(),
        "Remaining Rows": filtered.shape[0],
        "Rows Removed": original_df.shape[0] - filtered.shape[0]
    })

comparison_df = pd.DataFrame(comparison)

comparison_df

,Threshold,Remaining Diseases,Remaining Rows,Rows Removed
0,5,721,246823,122
1,10,677,246512,433
2,20,607,245548,1397


In [6]:
# ============================================================
# Step 7: Preview Dataset After Threshold = 10
# ============================================================

threshold = 10

disease_counts = original_df["diseases"].value_counts()

valid_diseases = disease_counts[disease_counts >= threshold].index

preview_df = original_df[original_df["diseases"].isin(valid_diseases)]

print("Preview Shape:", preview_df.shape)
print("Remaining Diseases:", preview_df["diseases"].nunique())

preview_df["diseases"].value_counts().describe()

Preview Shape: (246512, 378)
Remaining Diseases: 677


,count
count,677.000000
mean,364.124077
std,353.788719
min,10.000000
25%,62.000000
50%,242.000000
75%,607.000000
max,1219.000000


In [7]:
# ============================================================
# Step 8: Apply Rare Disease Filtering
# ============================================================

MIN_SAMPLES = 10

disease_counts = original_df["diseases"].value_counts()

valid_diseases = disease_counts[disease_counts >= MIN_SAMPLES].index

df = original_df[original_df["diseases"].isin(valid_diseases)].copy()

print("=" * 50)
print("Rare Disease Filtering Applied")
print("=" * 50)
print(f"Minimum Samples Required : {MIN_SAMPLES}")
print(f"Dataset Shape            : {df.shape}")
print(f"Remaining Diseases       : {df['diseases'].nunique()}")

Rare Disease Filtering Applied
Minimum Samples Required : 10
Dataset Shape            : (246512, 378)
Remaining Diseases       : 677


In [8]:
# ============================================================
# Step 9: Verify Dataset After Rare Disease Filtering
# ============================================================

print("=" * 60)
print("DATASET SUMMARY AFTER FILTERING")
print("=" * 60)

print(f"Rows                : {df.shape[0]:,}")
print(f"Columns             : {df.shape[1]}")
print(f"Remaining Diseases  : {df['diseases'].nunique()}")
print(f"Missing Values      : {df.isnull().sum().sum()}")
print(f"Duplicate Rows      : {df.duplicated().sum()}")

print("=" * 60)

DATASET SUMMARY AFTER FILTERING
Rows                : 246,512
Columns             : 378
Remaining Diseases  : 677
Missing Values      : 0
Duplicate Rows      : 57093


## Observation

- Diseases having fewer than **10 samples** were removed.
- Only **433 records** were removed from the original dataset.
- The dataset still contains **246,512 records** and **677 disease classes**.
- No missing values were introduced after filtering.
- The dataset retains more than **99.8%** of the original data while improving class reliability for machine learning.

In [9]:
# ============================================================
# Step 11: Separate Features and Target
# ============================================================

X = df.drop("diseases", axis=1)
y = df["diseases"]

print("=" * 50)
print("FEATURES AND TARGET CREATED")
print("=" * 50)
print("Feature Shape :", X.shape)
print("Target Shape  :", y.shape)

FEATURES AND TARGET CREATED
Feature Shape : (246512, 377)
Target Shape  : (246512,)


In [10]:
# ============================================================
# Step 12: Label Encoding
# ============================================================

label_encoder = LabelEncoder()

y_encoded = label_encoder.fit_transform(y)

print("=" * 50)
print("LABEL ENCODING COMPLETED")
print("=" * 50)
print("Total Classes :", len(label_encoder.classes_))

LABEL ENCODING COMPLETED
Total Classes : 677


In [11]:
# ============================================================
# Step 13: Train-Test Split
# ============================================================

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y_encoded,
    test_size=0.20,
    random_state=42,
    stratify=y_encoded
)

print("=" * 50)
print("TRAIN TEST SPLIT COMPLETED")
print("=" * 50)

print("Training Set :", X_train.shape)
print("Testing Set  :", X_test.shape)

TRAIN TEST SPLIT COMPLETED
Training Set : (197209, 377)
Testing Set  : (49303, 377)


In [12]:
print("=" * 50)
print("TRAINING LABELS :", len(set(y_train)))
print("TEST LABELS     :", len(set(y_test)))
print("=" * 50)

TRAINING LABELS : 677
TEST LABELS     : 677


In [14]:
# ============================================================
# Step 15: Save Processed Dataset
# ============================================================

processed_path = "/content/drive/MyDrive/AI-Powered-Healthcare-Diagnosis-Assistant/Processed_Data"

X_train.to_csv(f"{processed_path}/X_train.csv", index=False)
X_test.to_csv(f"{processed_path}/X_test.csv", index=False)

pd.DataFrame(y_train, columns=["Disease"]).to_csv(
    f"{processed_path}/y_train.csv", index=False
)

pd.DataFrame(y_test, columns=["Disease"]).to_csv(
    f"{processed_path}/y_test.csv", index=False
)

print("Processed datasets saved successfully.")

Processed datasets saved successfully.


In [15]:
import joblib

In [16]:
# ============================================================
# Step 17: Save Label Encoder
# ============================================================

joblib.dump(
    label_encoder,
    "/content/drive/MyDrive/AI-Powered-Healthcare-Diagnosis-Assistant/Processed_Data/label_encoder.pkl"
)

print("Label Encoder saved successfully.")

Label Encoder saved successfully.


In [17]:
# ============================================================
# Step 18: Save Feature Names
# ============================================================

joblib.dump(
    X.columns.tolist(),
    "/content/drive/MyDrive/AI-Powered-Healthcare-Diagnosis-Assistant/Processed_Data/feature_names.pkl"
)

print("Feature names saved successfully.")

Feature names saved successfully.


# Preprocessing Summary

## Tasks Completed

- Loaded the validated healthcare dataset.
- Created a backup of the original dataset.
- Compared rare disease thresholds (5, 10, and 20).
- Selected a minimum threshold of **10 samples per disease**.
- Removed rare disease classes.
- Separated features and target.
- Encoded disease labels using `LabelEncoder`.
- Performed stratified 80:20 train-test split.
- Exported processed datasets.
- Saved the label encoder.
- Saved the feature names for deployment.

## Conclusion

The dataset has been successfully preprocessed and is ready for machine learning model training.